esp32: Convert CONFIG_MP_NUM_CPUS handling

galak opened the pull request.

galak committed:
soc: xtensa: esp32: Convert CONFIG_MP_NUM_CPUS handling

Move runtime checks to use arch_num_cpus().  This is to allow
runtime determination of the number of CPUs in the future.

Signed-off-by: Kumar Gala <kumar.gala@intel.com>

galak committed:
interrupt_controller: intc_esp32: Convert CONFIG_MP_NUM_CPUS handling

Move runtime checks to use arch_num_cpus() and build checks
to use CONFIG_MP_MAX_NUM_CPUS.  This is to allow runtime
determination of the number of CPUs in the future.

Signed-off-by: Kumar Gala <kumar.gala@intel.com>

dcpleung commented on soc/xtensa/esp32/esp32-mp.c:
```diff
@@ -270,7 +270,7 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 
 	cpus_active[0] = true;
-	cpus_active[CONFIG_MP_NUM_CPUS - 1] = true;
+	cpus_active[cpu_num - 1] = true;
```
Shouldn't this be simply `[cpu_num]`? Since `cpu_num == 1` is the only valid number, this results in setting `cpus_active[0]` to true which is not the same behavior as before.

galak replied:
We get a compile error when I set this to `[cpu_num]`, since that can be outside of the array bounds.

galak replied:
Looking at it, this should be `cpus_active[cpu_num]`.  So I added ifdef protection to only build this code if CONFIG_SMP is enabled.

sylvioalves commented on drivers/interrupt_controller/intc_esp32.c:
```diff
@@ -260,7 +260,11 @@ int esp_intr_reserve(int intno, int cpu)
-	return (intr_alloc_table[intr * CONFIG_MP_NUM_CPUS + cpu].handler != default_intr_handler);
+	bool r;
+
+	r = intr_alloc_table[intr * CONFIG_MP_MAX_NUM_CPUS + cpu].handler != default_intr_handler;
```
Was there any compiling warning in here as well?

galak replied:
It was line length.  So split it out this way.

galak committed:
soc: xtensa: esp32: Add CONFIG_SMP protection

Add #ifdef CONFIG_SMP around code that only needs to exist when
SMP is enabled.

Also include ksched.h to get decleration of z_sched_ipi.

Signed-off-by: Kumar Gala <kumar.gala@intel.com>

andyross commented on soc/xtensa/esp32/esp32-mp.c:
```diff
@@ -39,9 +40,10 @@ struct cpustart_rec {
 volatile struct cpustart_rec *start_rec;
 static void *appcpu_top;
 static bool cpus_active[CONFIG_MP_MAX_NUM_CPUS];
+extern void z_sched_ipi(void);
```
This is defined in ksched.h, FWIW.  But checking the code I don't see it called in this C file (it isn't useful to SOC code), just remove.
